module/zfs/bpobj.c

```diff
@@ -264,97 +306,176 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 		dbuf = NULL;
 	}
 	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
+		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
+		    bpo->bpo_object, (i + 1) * sizeof (blkptr_t),
+		    DMU_OBJECT_END, tx));
 	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
+	return (err);
+}
 
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
+static void
+bpobj_iterate_cleanup(list_t *stack, int err)
+{
+	bpobj_t *bpo;
+
+	for (bpobj_info_t *bpi_cleanup = list_head(stack); bpi_cleanup != NULL;
+	    bpi_cleanup = list_next(stack, bpi_cleanup)) {
+		ASSERT3U(err, !=, 0);
+		bpo = bpi_cleanup->bpi_bpo;
+		ASSERT3P(bpo, !=, NULL);
+		if (bpi_cleanup->bpi_parent != NULL) {
+			bpobj_close(bpo);
+			kmem_free(bpi_cleanup->bpi_bpo, sizeof (bpobj_t));
+		}
+		kmem_free(bpi_cleanup, sizeof (bpobj_info_t));
 	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
+}
 
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
+#define	SUBOBJS(x) x->bpo_phys->bpo_num_subobjs
 
-		offset = i * sizeof (uint64_t);
-		blkoff = P2PHASE(i, epb);
+static int
+bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
+{
+	list_t stack;
+	bpobj_t *bpo = initial_bpo;
+	uint64_t freed, comp_freed, uncomp_freed;
+	dmu_buf_t *dbuf = NULL;
+	int err = 0;
+
+	list_create(&stack, sizeof (bpobj_info_t),
+	    offsetof(bpobj_info_t, bpi_node));
+	for (bpobj_info_t *bpi = bpi_alloc(bpo, NULL, 0, 0);
+	    bpi != NULL; bpi = list_remove_head(&stack)) {
+		bpo = bpi->bpi_bpo;
+		freed = comp_freed = uncomp_freed = 0;
+		mutex_enter(&bpo->bpo_lock);
+		ASSERT(bpobj_is_open(bpo));
+		if (free)
+			dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+
+		err = bpobj_free_blkptrs(bpi, func, arg, tx, free);
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			goto out;
+		}
+		/* A subobj with no subobjs below it can be cleaned up. */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			if (free && (bpi->bpi_parent != NULL)) {
+				bpobj_t *p = bpi->bpi_parent->bpi_bpo;
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
```

Comment:
it doesn't seem right to be accessing bpo_phys after we've freed the object it references.  I think we should do this after "eliminate the empty subobj".  And maybe we can move `ASSERT0(bpo->bpo_phys->bpo_bytes/comp/uncomp);` before this `if (free...` block?